File: solutions/haskell/bob/1/src/Bob.hs
{-# LANGUAGE OverloadedStrings #-}

module Bob (responseFor) where

import qualified Data.Text as T
import           Data.Text (Text)
import           Data.Char (isUpper, isLetter)

responseFor :: Text -> Text
responseFor xs 
    | vraag && all_hoofdletters = "Calm down, I know what I'm doing!"
    | vraag = "Sure."
    | all_hoofdletters = "Whoa, chill out!"
    | T.null $ T.strip xs = "Fine. Be that way!"
    | otherwise = "Whatever."
    where stripped = T.strip xs
          letters = T.filter isLetter stripped
          all_hoofdletters = T.length letters > 0 && T.all isUpper letters
          vraag = T.isSuffixOf "?" stripped

File: solutions/java/bob/1/src/main/java/Bob.java
class Bob {

    String hey(String input) {
        input = input.trim();

        if (input.isEmpty()) {
            return "Fine. Be that way!";
        }

        boolean question = isQuestion(input);
        boolean yell = isYelling(input);

        if (question && yell) {
            return "Calm down, I know what I'm doing!";
        } else if (question) {
            return "Sure.";
        } else if (yell) {
            return "Whoa, chill out!";
        } else {
            return "Whatever.";
        }
    }

    private static boolean isQuestion(String input) {
        return input.endsWith("?");
    }

    private boolean isYelling(String input) {
        boolean hasLetter = false;

        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);
            if (Character.isLetter(c)) {
                if (Character.isLowerCase(c)) {
                    return false;
                }
                hasLetter = true;
            }
        }

        return hasLetter;
    }
}